include/misc/libc-hooks.h

```diff
@@ -27,16 +27,27 @@ __syscall int _zephyr_read(char *buf, int nbytes);
 
 __syscall int _zephyr_write(const void *buf, int nbytes);
 
+#ifdef CONFIG_APP_SHARED_MEM
+/* Memory partition containing newlib's globals. This includes all the globals
+ * within libc.a and the supporting zephyr hooks, but not the malloc arena.
+ */
+extern struct k_mem_partition z_newlib_partition;
+#endif /* CONFIG_APP_SHARED_MEM */
+
 #else
 /* Minimal libc */
 
 __syscall int _zephyr_fputc(int c, FILE *stream);
 
 __syscall size_t _zephyr_fwrite(const void *_MLIBC_RESTRICT ptr, size_t size,
 				size_t nitems, FILE *_MLIBC_RESTRICT stream);
-
 #endif /* CONFIG_NEWLIB_LIBC */
 
+#ifdef CONFIG_APP_SHARED_MEM
+/* Memory partition containing the libc malloc arena */
+extern struct k_mem_partition z_malloc_partition;
```

Comment:
I have a couple of questions regarding z_malloc_partition:
First, it looks like this is an application domain partition; could that be reflected in the variable's name?
Second, the actual partition is defined in malloc.c and in newlib (lib-hooks.c) in two different ways (in the first with the APPMEM_PARTITION_DEFINE and in second with explicit assignments of the struct properties) . Is this intentional, or could this be unified?

Reply:
> application domain partition; could that be reflected in the variable's name?

To me these are all just memory domain partitions. If you need the malloc arena, you add this to your domain. I think the "app shared memory" subsystem is poorly named, what we are really doing is just tying in the definition of k_mem_partitions with linker magic. I have an urge to delete the Kconfig for it entirely and just make this a core part of memory domains, see #13075 

> Is this intentional, or could this be unified?

Newlib is a special case, as for non power of two arches the newlib malloc arena is all the unused memory at the end of SRAM, we don't specifically declare a byte buffer for it, instead it's just linker symbols. There's nothing to tag with Z_APP_BMEM so I populate the k_mem_partition at runtime via an init function.

malloc_prepare should be unnecessary for power of two arches, there should instead be a Z_APP_BMEM tag on heap_base array instead, I'll change this.

Reply:
Updated.